File: 01-Arrays/20250730-intermediate/ts/solution.ts
// TODO: Implement the mergeSortedArrays function
// This function should merge two pre-sorted arrays into a single sorted array
// Requirements:
// - Take two sorted arrays as input
// - Return a new array containing all elements in sorted order
// - Handle different array sizes
// - Preserve the sorted order property
// - Use O(m+n) time complexity where m and n are array lengths
export function mergeSortedArrays(arr1: number[], arr2: number[]): number[] {
    // TODO: Implement merge logic using two-pointer technique;
    // 1. Create result array with capacity arr1.length + arr2.length
    // 2. Use two pointers to traverse both arrays simultaneously
    // 3. Compare elements and add the smaller one to result
    // 4. Handle remaining elements when one array is exhausted
    const out = []
    let [c1, c2] = [0, 0]

    while (c1 < arr1.length && c2 < arr2.length) {
        if (arr1[c1] <= arr2[c2]) {
            out.push(arr1[c1])
            c1++
        } else {
            out.push(arr2[c2])
            c2++
        }
    }
    return out.concat(arr1.slice(c1), arr2.slice(c2))

}

File: 01-Arrays/20250730-intermediate/go/solution.go
package main


// TODO: Implement the MergeSortedArrays function
// This function should merge two pre-sorted arrays into a single sorted array
// Requirements:
// - Take two sorted arrays as input
// - Return a new array containing all elements in sorted order
// - Handle different array sizes
// - Preserve the sorted order property
// - Use O(m+n) time complexity where m and n are array lengths
func MergeSortedArrays(arr1, arr2 []int) []int {
	// TODO: Implement merge logic using two-pointer technique
	n := len(arr1) + len(arr2)
	out := make([]int, n)
	c1 := 0
	c2 := 0
	for i := 0; i < n; i++ {
		if c1 == len(arr1) {
			for c2 != len(arr2) {
				out[i] = arr2[c2]
				i++
				c2++
			}
			break
		}
		if c2 == len(arr2) {
			for c1 != len(arr1) {
				out[i] = arr1[c1]
				i++
				c1++
			}
			break
		}
		if arr1[c1] < arr2[c2] {
			out[i] = arr1[c1]
			c1++
		} else {
			out[i] = arr2[c2]
			c2++
		}

	}
	return out
}

// Industry standard solution using cleaner two-pointer approach
func MergeSortedArraysStandard(arr1, arr2 []int) []int {
	result := make([]int, 0, len(arr1)+len(arr2))
	i, j := 0, 0
	
	// Merge while both arrays have elements
	for i < len(arr1) && j < len(arr2) {
		if arr1[i] <= arr2[j] {
			result = append(result, arr1[i])
			i++
		} else {
			result = append(result, arr2[j])
			j++
		}
	}
	
	// Append remaining elements
	result = append(result, arr1[i:]...)
	result = append(result, arr2[j:]...)
	
	return result
}

